### DISCUSSION TITLE
[BUG] `zsh: segmentation fault`

### DISCUSSION THREAD

**Discussion by qusoulspace** [dashingsoft/pyarmor#1329] [Q&A]
I was trying to obfuscate the main.py file which contains print('Hello World'). After obfuscation, the script runs and outputs everything, but when I package via pyinstaller I get the error zsh: segmentation fault. If you do not obfuscate the same script, then everything is OK

```
$ pyarmor gen main.py
INFO     Python 3.11.3
INFO     Pyarmor 8.2.2 (trial), 000000, non-profits
INFO     Platform darwin.x86_64
INFO     search inputs ...
INFO     find script main.py
INFO     find 1 top resources
INFO     start to generate runtime files
INFO     target platforms {'darwin.x86_64'}
INFO     write dist/pyarmor_runtime_000000/pyarmor_runtime.so
INFO     patch runtime file
INFO     codesign -f -s - --all-architectures --timestamp dist/pyarmor_runtime_000000/pyarmor_runtime.so
INFO     generate runtime files OK
INFO     start to obfuscate scripts
INFO     process resource "main"
INFO     obfuscating file main.py
INFO     write dist/main.py
INFO     obfuscate scripts OK

```
then packing with pyinstaller
```
$ pyinstaller --onefile main.py
327 INFO: PyInstaller: 5.12.0
327 INFO: Python: 3.11.3
345 INFO: Platform: macOS-13.4-x86_64-i386-64bit
346 INFO: wrote /Users/User/Desktop/py/dist/main.spec
352 INFO: Extending PYTHONPATH with paths
['/Users/User/Desktop/py/dist']
629 INFO: checking Analysis
632 INFO: Building because /Users/User/Desktop/py/dist/main.py changed
632 INFO: Initializing module dependency graph...
635 INFO: Caching module graph hooks...
648 INFO: Analyzing base_library.zip ...
2048 INFO: Loading module hook 'hook-heapq.py' from '/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/PyInstaller/hooks'...
2152 INFO: Loading module hook 'hook-encodings.py' from '/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/PyInstaller/hooks'...
3782 INFO: Loading module hook 'hook-pickle.py' from '/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/PyInstaller/hooks'...
5781 INFO: Caching module dependency graph...
5893 INFO: running Analysis Analysis-00.toc
5909 INFO: Analyzing /Users/User/Desktop/py/dist/main.py
5922 INFO: Processing module hooks...
5932 INFO: Looking for ctypes DLLs
5943 INFO: Analyzing run-time hooks ...
5945 INFO: Including run-time hook '/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/PyInstaller/hooks/rthooks/pyi_rth_inspect.py'
5950 INFO: Looking for dynamic libraries
6543 INFO: Looking for eggs
6544 INFO: Using Python library /Library/Frameworks/Python.framework/Versions/3.11/Python
6548 INFO: Warnings written to /Users/User/Desktop/py/dist/build/main/warn-main.txt
6559 INFO: Graph cross-reference written to /Users/User/Desktop/py/dist/build/main/xref-main.html
6566 INFO: checking PYZ
6568 INFO: Building because /Users/User/Desktop/py/dist/pyarmor_runtime_000000/__init__.py changed
6568 INFO: Building PYZ (ZlibArchive) /Users/User/Desktop/py/dist/build/main/PYZ-00.pyz
6796 INFO: Building PYZ (ZlibArchive) /Users/User/Desktop/py/dist/build/main/PYZ-00.pyz completed successfully.
6804 INFO: EXE target arch: x86_64
6804 INFO: Code signing identity: None
6805 INFO: checking PKG
6806 INFO: Building because toc changed
6806 INFO: Building PKG (CArchive) main.pkg
12268 INFO: Building PKG (CArchive) main.pkg completed successfully.
12270 INFO: Bootloader /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/PyInstaller/bootloader/Darwin-64bit/run
12270 INFO: checking EXE
12271 INFO: Building because toc changed
12271 INFO: Building EXE from EXE-00.toc
12281 INFO: Copying bootloader EXE to /Users/User/Desktop/py/dist/dist/main
12282 INFO: Converting EXE to target arch (x86_64)
12302 INFO: Removing signature(s) from EXE
12321 INFO: Appending PKG archive to EXE
12414 INFO: Fixing EXE headers for code signing
12427 INFO: Re-signing the EXE
12561 INFO: Building EXE from EXE-00.toc completed successfully.
```

**Comment by jondy** [comment]
Please first upgrade Pyarmor to latest version 8.2.8
Then check the solutions in the documentation questions and topics.

**Reply by qusoulspace** [reply]
I updated to new version same error.
pyarmor gen main.py
INFO     Python 3.11.3
INFO     Pyarmor 8.2.8 (trial), 000000, non-profits
INFO     Platform darwin.x86_64
INFO     search inputs ...
INFO     find script main.py
INFO     find 1 top resources
INFO     start to generate runtime files
INFO     target platforms {'darwin.x86_64'}
INFO     write dist/pyarmor_runtime_000000/pyarmor_runtime.so
INFO     patch runtime file
INFO     codesign -f -s - --all-architectures --timestamp dist/pyarmor_runtime_000000/pyarmor_runtime.so
INFO     generate runtime files OK
INFO     start to obfuscate scripts
INFO     process resource "main"
INFO     obfuscating file main.py
INFO     write dist/main.py
INFO     obfuscate scripts OK

**Comment by jondy** [comment]
Here are solutions for apple segmentfault
https://pyarmor.readthedocs.io/en/v8.2.8/questions.html#segment-fault-in-apple

**Comment by Warfront1** [comment]
@qusoulspace were you ever able to find a solution for this?
I am running into the same problem on 8.2.9 on an M1 Mac Mini.
I am able to run my obfuscated perfectly, but when I try to run the standalone executable generated by p2app I get a segmentfault.
`python ./dist/obf/python/setup.py py2app`
The provided pyarmor documentation workarounds don't seem to be working.

**Comment by eranegozy** [comment]
Hi. Similar issue,  but I get a seg fault when I run `pyarmor gen`. I tried the codesign fix, but that did not work. I am on an M2 mac (OS 13.4.1). Here is the output:

```
(fmp) eran@Lightbulb fmplib % pyarmor gen unit1.py                                                                                   
INFO     Python 3.11.4
INFO     Pyarmor 8.2.9 (trial), 000000, non-profits
INFO     Platform darwin.aarch64
INFO     search inputs ...
INFO     find script unit1.py
INFO     find 1 top resources
zsh: segmentation fault  pyarmor gen unit1.py
(fmp) eran@Lightbulb fmplib % codesign -s - -f ~/miniconda_arm/envs/fmp/lib/python3.11/site-packages/pyarmor/cli/core/pytransform3.so
/Users/eran/miniconda_arm/envs/fmp/lib/python3.11/site-packages/pyarmor/cli/core/pytransform3.so: replacing existing signature
(fmp) eran@Lightbulb fmplib % pyarmor gen unit1.py                                                                                   
INFO     Python 3.11.4
INFO     Pyarmor 8.2.9 (trial), 000000, non-profits
INFO     Platform darwin.aarch64
INFO     search inputs ...
INFO     find script unit1.py
INFO     find 1 top resources
zsh: segmentation fault  pyarmor gen unit1.py
```

Any ideas on what I should try next?

**Comment by Warfront1** [comment]
I may write up a bug report later, but there are problems with the latest Pyarmor, Pyinstaller, and the latest Python (3.11.4 at the time of writing).
There are breaking issues in Pyarmor depending on Pyinstaller methods that are no longer available in the latest version (5.13.0)
My recommendation is to use the following versions: pyinstaller 5.11.0, pyarmor 8.2.9, and Python 3.11.2 (I think Python 3.11.3 may work as well).

Also, I gave up on py2app and switched my entire packaging solution to Pyinstaller.
The manual instructions for `Packing obfuscated scripts manually` with Pyarmor did not work.
However, I was able to get the high level pack command to work based on the documentation though.
https://pyarmor.readthedocs.io/en/stable/topic/repack.html?highlight=pack#insight-into-pack-command

**Comment by jondy** [comment]
@Warfront1 you're right. Pyarmor doesn't full test with every PyInstaller version, it may not work with some PyInstaller versions.

That's why I document how to pack the obfuscated scripts manually.

`repack` is an extra feature in Pyarmor, note core feature.

And it's better to upgrade Pyarmor to 8.3.0,  the extension `pyarmor_runtime.so` doesn't bind to fixed libpythonX.Y.lib, but find it automatically. It may solve many issues in Darwin.
